tc/core/polyhedral/memory_promotion.cc

```diff
@@ -314,11 +360,24 @@ isl::multi_aff TensorReferenceGroup::promotion() const {
   isl::map map = scopedAccesses();
   auto accessSpace = map.get_space();
 
-  // lower bounds space is S -> P; which we transform into [S -> O] -> P
-  auto lowerBounds = approximation.lowerBounds().pullback(
-      isl::multi_aff::domain_map(accessSpace));
-  auto promotion = isl::multi_aff::range_map(accessSpace) - lowerBounds;
-  return promotion;
+  // Construct a projection multi-aff in [S -> O] -> S
+  // for further precomposition.
+  auto originalSpaceInserter = isl::multi_aff::domain_map(accessSpace);
+
+  // Lower bounds and offsets space is S -> O; transform into [S -> O] -> O.
+  auto lowerBounds =
+      approximation.lowerBounds().pullback(originalSpaceInserter);
+  auto offsets = approximation.strideOffsets.pullback(originalSpaceInserter);
+
+  // Create promotion starting by identity in [S -> O] -> O.
+  auto original = isl::multi_aff::range_map(accessSpace);
+  auto promotion =
+      (original - offsets) / approximation.strideValues - lowerBounds;
+
+  // The target space is that of the promoted tensor, but the group name is not
+  // yet known.  Just reset the tuple id of the target space to obtain
+  // [S -> O] -> P where P has empty tuple id.
```

Comment:
What's the harm in keeping the original name until the group name is known?

Reply:
I'd argue that the spaces are conceptually different (array in global memory vs array in some other memory), so they should not have matching tuples.  OTOH, using an unnamed tuple makes it look like it's the same space for all groups.

Previous implementation did not have the tuple name for the "promoted" space, hence the comment about lower bounds living in "S -> P", which is no longer the case.